oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -0,0 +1,147 @@
+package com.google.auth.credentialaccessboundary;
+
+import static com.google.auth.oauth2.OAuth2Credentials.getFromServiceLoader;
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auth.Credentials;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.OAuth2Utils;
+import com.google.auth.oauth2.StsRequestHandler;
+import com.google.auth.oauth2.StsTokenExchangeRequest;
+import com.google.auth.oauth2.StsTokenExchangeResponse;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.io.IOException;
+
+public final class ClientSideCredentialAccessBoundaryFactory {
+  private final GoogleCredentials sourceCredential;
+  private final transient HttpTransportFactory transportFactory;
+  private final String tokenExchangeEndpoint;
+  private String acceessBoundarySessionKey;
+  private AccessToken intermediaryAccessToken;
+
+  private ClientSideCredentialAccessBoundaryFactory(Builder builder) {
+    this.transportFactory =
+        firstNonNull(
+            builder.transportFactory,
+            getFromServiceLoader(HttpTransportFactory.class, OAuth2Utils.HTTP_TRANSPORT_FACTORY));
+    this.sourceCredential = checkNotNull(builder.sourceCredential);
+
+    // Default to GDU when not supplied.
+    String universeDomain;
+    if (builder.universeDomain == null || builder.universeDomain.trim().isEmpty()) {
+      universeDomain = Credentials.GOOGLE_DEFAULT_UNIVERSE;
+    } else {
+      universeDomain = builder.universeDomain;
+    }
+
+    // Ensure source credential's universe domain matches.
+    try {
+      if (!universeDomain.equals(sourceCredential.getUniverseDomain())) {
+        throw new IllegalArgumentException(
+            "The client side access boundary credential's universe domain must be the same as the source "
+                + "credential.");
+      }
+    } catch (IOException e) {
+      // Throwing an IOException would be a breaking change, so wrap it here.
+      throw new IllegalStateException(
+          "Error occurred when attempting to retrieve source credential universe domain.", e);
+    }
+    String TOKEN_EXCHANGE_URL_FORMAT = "https://sts.{universe_domain}/v1/token";
+    this.tokenExchangeEndpoint =
+        TOKEN_EXCHANGE_URL_FORMAT.replace("{universe_domain}", universeDomain);
+  }
+
+  public void fetchCredentials() throws IOException {
+    try {
+      this.sourceCredential.refreshIfExpired();
+    } catch (IOException e) {
+      throw new IOException("Unable to refresh the provided source credential.", e);
```

Comment:
qq, why catch and rethrow this IOException? Is the message from the earlier IOException not detailed enough?

Also, can you add Javadocs for this public method?

Reply:
Good question. I saw this was the way it was handled in `DownscopedCredentials.java` and assumed that was necessary. @lsirac might know more about it.
 https://github.com/googleapis/google-auth-library-java/blob/7c5ed2fb2b1f6ddc74f3ec4873d2915f58d7128f/oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java#L137